File: medium/cpp/654.cpp
// https://leetcode.com/problems/maximum-binary-tree/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    TreeNode* constructMaximumBinaryTree(std::vector<int>& nums)
    {
        return solve(nums, 0, std::size(nums));
    }
    
private:
    TreeNode* solve(const std::vector<int>& nums, int left, int right)
    {
        auto maxPos = left;
        auto maxNum = nums[left];
        
        for (int pos = left; pos < right; pos++)
        {
            if (nums[pos] > maxNum)
            {
                maxNum = nums[pos];
                maxPos = pos;
            }
        }
        
        TreeNode* node = new TreeNode(maxNum);
        
        if (left < maxPos)
        {
            node->left = solve(nums, left, maxPos);
        }
        
        if (maxPos + 1 < right)
        {
            node->right = solve(nums, maxPos + 1, right);
        }
        
        return node;
    }
};

File: medium/python/654.py
# https://leetcode.com/problems/maximum-binary-tree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> Optional[TreeNode]:
        return self.build(nums, 0, len(nums))

    def build(self, nums, left, right):
        if left == right:
            return None

        idx = left
        for i in range(left, right):
            if nums[i] > nums[idx]:
                idx = i

        return TreeNode(nums[idx], self.build(nums, left, idx), self.build(nums, idx + 1, right))